beluga/test/beluga/estimation/test_cluster_based_estimation.cpp

```diff
@@ -0,0 +1,373 @@
+
+  // test proper
+  auto tested_map = original_map;
+  cap_grid_cell_data_weights(tested_map, weight_cap);
+
+  for (const auto& [hash, grid_cell] : tested_map) {
+    const auto original_grid_cell = original_map[hash];
+    EXPECT_DOUBLE_EQ(grid_cell.weight, std::min(original_grid_cell.weight, weight_cap));
+  }
+}
+
+TEST_F(ClusterBasedEstimationDetailTesting, PriorityQueuePopulationStep) {
+  // data preparation
+  auto map = generate_test_grid_cell_data_map();
+  const auto weight_cap = calculate_percentile_weight_threshold(map, 0.9);
+  cap_grid_cell_data_weights(map, weight_cap);
+
+  // test proper
+  auto prio_queue = populate_priority_queue(map);
+  EXPECT_EQ(prio_queue.size(), map.size());
+
+  // the top element necessarily has the highest weight equal to the cap
+  EXPECT_DOUBLE_EQ(prio_queue.top().priority, weight_cap);
+
+  // from there on the weights should be strictly decreasing
+  auto prev_weight = prio_queue.top().priority;
+  while (!prio_queue.empty()) {
+    const auto top = prio_queue.top();
+    EXPECT_GE(prev_weight, top.priority);
+    prev_weight = top.priority;
+    prio_queue.pop();
+  }
+}
+
+TEST_F(ClusterBasedEstimationDetailTesting, MapGridCellsToClustersStep) {
+  // create a map with four independent peaks
+  std::vector<std::tuple<double, double, double>> coordinates;
+  for (double x = 0.0; x < 36.0; x += 1.0) {
+    for (double y = 0.0; y < 36.0; y += 1.0) {
+      const auto weight = std::abs(std::sin(10.0 * x * M_PI / 180.0)) * std::abs(std::sin(10.0 * y * M_PI / 180.0));
+      coordinates.emplace_back(x, y, weight);
+    }
+  }
+
+  GridCellDataMap2D map;
+  for (const auto& [x, y, w] : coordinates) {
+    const auto state = SE2d{SO2d{0.}, Vector2d{x, y}};
+    map.emplace(spatial_hash_function(state), GridCellData{w, 0, state, std::nullopt});
+  }
+
+  const auto weight_cap = calculate_percentile_weight_threshold(map, 0.9);
+
+  cap_grid_cell_data_weights(map, weight_cap);
+
+  const auto adjacent_grid_cells = {
+      SE2d{SO2d{0.0}, Vector2d{+kSpatialHashResolution, 0.0}},
+      SE2d{SO2d{0.0}, Vector2d{-kSpatialHashResolution, 0.0}},
+      SE2d{SO2d{0.0}, Vector2d{0.0, +kSpatialHashResolution}},
+      SE2d{SO2d{0.0}, Vector2d{0.0, -kSpatialHashResolution}},
+  };
+
+  // test proper
+
+  // only cells beyond the 10% weight percentile to avoid the messy border
+  // between clusters beneath that threshold
+  const auto ten_percent_threshold = calculate_percentile_weight_threshold(map, 0.15);
+  map_cells_to_clusters(map, spatial_hash_function, adjacent_grid_cells, weight_cap);
+
+  auto cells_above_minimum_threshold_view =
+      coordinates | ranges::views::filter([&](const auto& c) { return std::get<2>(c) >= ten_percent_threshold; });
+
+  auto quadrant_1_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) < 18.0 && std::get<1>(c) < 18.0;
```

Comment:
@glpuga nit: using structured bindings may help readability here. Also, named constants for grid width and height.